Graph types (#1391)
neo4j/docs-cypher: modules/ROOT/pages/schema/graph-types/set-graph-types.adoc

----

[NOTE]
Properties specified in a `REQUIRE` clause are not required to exist on the node or relationship element types to which the constraint applies.
For example, the following is a valid element type key constraint defined against a node element type: `(n:Label \=> {property1 :: STRING, property2 :: INTEGER}) REQUIRE (n.property3, n.property4) IS KEY`.

The `REQUIRE` keyword can also be used as an alternative syntax to define several property uniqueness and key constraints within the same node or relationship element type.

.Syntax for defining multiple property uniqueness or key constraints using the `REQUIRE` keyword
----
(n:Label => “{“property1 :: <TYPE>, property2 :: <TYPE>”}”)
    REQUIRE n.property1 IS {UNIQUE | KEY}
    REQUIRE n.property2 IS {UNIQUE | KEY}

()-”[“r:REL_TYPE => “{“r.property1 :: <TYPE>, r.property2 :: <TYPE> “}” “]”->()
    REQUIRE r.property1 IS {UNIQUE | KEY}
    REQUIRE r.property2 IS {UNIQUE | KEY}
----

[[naming-property-uniqueness-and-key-constraints]]
=== Naming property uniqueness and key constraints

Property uniqueness and key constraints can also be created using the `CONSTRAINT … FOR … REQUIRE` syntax.

> If so, they can be given a name which can be used in the `ALTER CURRENT GRAPH TYPE DROP` or `DROP CONSTRAINT` commands (if a name is not explicitly given, a name will be assigned).

Reviewer: might be good to mention here that the given names need to be unique similarly to how we do on the constraint page:
`The given name must be unique among both indexes and constraints.`

When situations like the example above arise, users should migrate the existing data to the new desired model and then apply the new graph type that will keep the data consistent from then on.

[[constraints-on-non-identifying-labels-and-types]]
=== Constraints on non-identifying node labels and relationship types

The examples thus far have included node and relationship element types, as well as property uniqueness and key constraints defined on identifying node labels and relationship types.

It is also possible to include constraints in a graph type on implied node labels, or on node labels and relationship types which are independent of a node or relationship element type, using the `CONSTRAINT … FOR … REQUIRE` syntax outlined xref:schema/graph-types/set-graph-types.adoc#naming-property-uniqueness-and-key-constraints[above].
This can be a useful tool if the data model requires certain attributes for nodes or relationships with non-identifying labels or types.

[NOTE]

> Constraints on non-identifying node labels and relationship types can be explicitly named when defined.This can be useful when see xref:schema/graph-types/drop-graph-type-elements.adoc#drop-constraints-on-identifying-and-non-identifying-labels-and-types[Drop constraints on identifying and non-identifying node labels and relationship types] (if a name is not explicitly given, a name will be assigned).

Reviewer: might be good to mention here that the given names need to be unique similarly to how we do on the constraint page:
`The given name must be unique among both indexes and constraints.`

:page-role: cypher-25-only new-neo4j-2025.xx enterprise-edition
= Set graph types

A graph type defines the schema of a database and is set with the command `ALTER CURRENT GRAPH TYPE SET`.
*Note that any previously defined graph types or xref:schema/constraints/index.adoc[constraints] are overwritten and replaced when a graph type is set.*

Graph types can include:

* xref:schema/graph-types/set-graph-types.adoc#node-element-types[Node element types]
* xref:schema/graph-types/set-graph-types.adoc#relationship-element-types[Relationship element types]
* xref:schema/graph-types/set-graph-types.adoc#constraints-on-identifying-labels-and-types[Property uniqueness and key constraints defined on identifying node labels or relationship types]

These categories are discussed in the section xref:schema/graph-types/set-graph-types.adoc#set-graph-type-empty-database[Set a graph type on an empty database].

Graph types can also include xref:schema/graph-types/set-graph-types.adoc#constraints-on-non-identifying-labels-and-types[property existence, property type, property uniqueness, and key constraints] defined on non-identifying node labels or relationship types.

> These are discussed in the section xref:schema/graph-types/set-graph-types.adoc#set-graph-type-populated-database[Set a graph type on a populated database].

Reviewer: Reading this (again I assume :P) feels a bit odd, first you list three things it includes in a way that make it seem it can only have those and then you have a btw paragraph that they can also include something else 🤔 
<img width="1600" height="566" alt="image" src="https://github.com/user-attachments/assets/69a63cae-67c4-4818-b560-24ecf03c53e2" />

Maybe better to do something like
```suggestion
Graph types can include:

* xref:schema/graph-types/set-graph-types.adoc#node-element-types[Node element types]
* xref:schema/graph-types/set-graph-types.adoc#relationship-element-types[Relationship element types]
* xref:schema/graph-types/set-graph-types.adoc#constraints-on-identifying-labels-and-types[Property uniqueness and key constraints defined on identifying node labels or relationship types]
* xref:schema/graph-types/set-graph-types.adoc#constraints-on-non-identifying-labels-and-types[property existence, property type, property uniqueness, and key constraints defined on non-identifying node labels or relationship types]

The three first categories are discussed in the section xref:schema/graph-types/set-graph-types.adoc#set-graph-type-empty-database[Set a graph type on an empty database]. The last category is discussed in the section xref:schema/graph-types/set-graph-types.adoc#set-graph-type-populated-database[Set a graph type on a populated database].
```

[NOTE]
====
Setting a graph type requires the following privileges:

> *  link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/database-administration/#access-control-database-administration-constraints[`CREATE CONSTRAINT`]

Reviewer: ```suggestion
* link:{neo4j-docs-base-uri}/operations-manual/current/authentication-authorization/database-administration/#access-control-database-administration-constraints[`CREATE CONSTRAINT`]
```

Reviewer: this wasn't actually fixed :P

Persons and pets are both classified as residents, and the data model captures how long both have lived in a particular city.

image::graph-types-data-model.svg[Data model showing persons and pets and the city they live in, width=700,role=popup]

[[set-graph-type-empty-database]]
== Set a graph type on an empty database

All databases are created with an empty graph type.
Since setting a graph type can be a more involved process if a database already has pre-existing data, it is recommended to set the desired graph type before populating a graph with data.

.Set a graph type to ensure the integrity of the data model in a database
[source, cypher]
----
ALTER CURRENT GRAPH TYPE SET {
    (:Person => :Resident {name :: STRING NOT NULL}),
    (:Pet => :Resident&Animal {healthCertificate :: STRING, name :: STRING}),
    (:City => {name :: STRING NOT NULL, population :: INTEGER}),
    (:Resident)-[:LIVES_IN => {since :: DATE NOT NULL}]->(:City)
}
----

[[node-element-types]]
=== Node element types

.Node element type example
image::graph-types-node-element-type.svg[Image of an example node element type, width=600,role=popup]

Node element types are defined by an *identifying label*.

> Node element types ensure that or more of the following is true for all nodes with an identifying label:

Reviewer: ```suggestion
Node element types ensure that one or more of the following is true for all nodes with an identifying label:
```

.Node element types in graph type
[source]
----
(:Person => :Resident {name :: STRING NOT NULL}),  // <1>
(:Pet => :Resident&Animal {healthCertificate :: STRING, name :: STRING}), // <2>
(:City => {name :: STRING NOT NULL, population :: INTEGER}) // <3>
----

<1> Node element type ensuring that nodes with a `Person` label must also have a `Resident` label and that `Person` nodes must have a `name` property of type `STRING`.
The `Person` label is the identifying label of this node element type, and `Resident` is the implied label.
<2> Node element type ensuring that nodes with a `Pet` label must also have a `Resident` and an `Animal` label.
The node element type also ensures that any `healthCertificate` and `name` properties on `Pet` nodes must be of type `STRING`.
The `Pet` label is the identifying label of this node element type while `Resident` and `Animal` are implied labels.
<3> Node element type ensuring nodes with a `City` label must have a `name` property of type `STRING`, and that any `population` properties are of type `INTEGER`.
`City` is the identifying label of this node element type, and there are no implied labels.

.Node element type syntax
[source, syntax]
----

> (:IdentifyingLabel => [:ImpliedLabel1[&...]]  [“{“property1 :: <TYPE> [NOT NULL][, ...]”}”])

Reviewer: same as we had on the syntax page
```suggestion
(:IdentifyingLabel => [:ImpliedLabel1[&...]]  [“{“ {property1 :: <TYPE> [NOT NULL]}[, ...]”}”])
```

I also assume we will see this on the rel elem type one, and maybe other syntax snippets but I won't add multiple comments on this to save your comment amount :P

----

If `ANY` is specified as the `<TYPE>`, the node element type will allow all valid property types to be assigned to the given property (`ANY` is only allowed if the property’s existence is enforced, that is if it is appended with `NOT NULL`).

[[relationship-element-types]]
=== Relationship element types

.Relationship element type example
image::graph-types-relationship-element-type.svg[Image of an example relationship element type, width=600,role=popup]

> Relationship element types are defined by an *identifying type*.

Reviewer: should it technically be
```suggestion
Relationship element types are defined by an *identifying relationship type*.
```

Author: We don't say *indentifying node label* for node element type so I think it is fine, especially given the above picture

Reviewer: But we also in general call them labels and relationship types 🤷  but we are saying both type and relationship type intermixed throughout the chapter so 🤷

====

[[identifying-node-labels-and-relationship-types]]
=== Identifying node labels and relationship types

Node element types have identifying node labels and relationship element types have identifying relationship types.
These identifying node labels and relationship types are used to uniquely identify a node or relationship element type in a graph type.
For example, in a node element type definition like `(:Person \=> :Resident {\...})`, the `Person` label is the identifying label, indicating that the node element type represents a node with a `Person` label.

*You cannot have different element types with the same identifying node label or relationship type.*

.Allowed
[source]
----
(:Pet => :Resident&Animal {healthCertificate :: STRING, name :: STRING}),
(:Person)-[:LIVES_IN => {since :: DATE NOT NULL}]->(:City),
(:Person)-[:BORN_IN => {dob :: DATE NOT NULL})]->(:City)
----

.Not allowed -- identifying node labels among node element types must be unique
[source]
----
(:Person => :Resident {name :: STRING}),
(:Person => :Citizen {name :: STRING})
----

> .Not allowed -- identifying relationship types among relationship element types must be unique

Reviewer: I would probably switch places on these two examples, since the second not-allowed one is connected to the allowed one it feels odd to have this one in-between them 🤷

The `REQUIRE` keyword can also be used as an alternative syntax to define several property uniqueness and key constraints within the same node or relationship element type.

.Syntax for defining multiple property uniqueness or key constraints using the `REQUIRE` keyword
----
(n:Label => “{“property1 :: <TYPE>, property2 :: <TYPE>”}”)
    REQUIRE n.property1 IS {UNIQUE | KEY}
    REQUIRE n.property2 IS {UNIQUE | KEY}

()-”[“r:REL_TYPE => “{“r.property1 :: <TYPE>, r.property2 :: <TYPE> “}” “]”->()
    REQUIRE r.property1 IS {UNIQUE | KEY}
    REQUIRE r.property2 IS {UNIQUE | KEY}
----

[[naming-property-uniqueness-and-key-constraints]]
=== Naming property uniqueness and key constraints

Property uniqueness and key constraints can also be created using the `CONSTRAINT … FOR … REQUIRE` syntax.
If so, they can be given a name which can be used in the `ALTER CURRENT GRAPH TYPE DROP` or `DROP CONSTRAINT` commands (if a name is not explicitly given, a name will be assigned).
The given name must be unique among both indexes and constraints.

> For more information, see xref:schema/graph-types/drop-graph-type-elements.adoc#drop-constraints-on-identifying-and-non-identifying-labels-and-types[Drop constraints on identifying and non-identifying node labels and relationship types].

Reviewer: I don't know what order I want these sentences in, as it is now the `For more information, ...` sounds like it's connected to the info about the name needing to be unique which isn't correct. And if I switch the last two sentences the sentence about unique names seems out of place :(

I guess the issue is that both the second and third sentence is connected to the first one and I'm not sure what to do about it

Author: Have rephrased.

CREATE (:Person:Resident {name: 'Alice Carlyle', ssn: 'HNB48182'}); // <4>
----

<1> This will fail because it would create a node with properties identical to an existing one, thereby violating the key constraint defined in the `Person` node element type.
<2> This will fail because the `Person` node element type requires nodes with a `Person` label to also have a `Resident` label.
<3> This will fail because not all properties constrained in the graph type key constraint (`name`, `ssn`) defined in the `Person` node element type are present.
<4> This will fail because a type-constrained property (`ssn`) is of a different type (`STRING`) than the one defined for the property in the `Person` node element type (`INTEGER`).

.Create data violating a relationship element type
[source, cypher, role=test-fail]
----
MATCH (alice:Person:Resident {name: 'Alice Carlyle'})
CREATE (:Human {name: 'Carl Ericsson' ssn: 765498321})-[:LIVES_IN {since: date('1999-10-17’)}]->(:City {name: 'Toronto'}); // <1>
CREATE (alice)-[:LIVES_IN {since: date('1999-10-17')}]->(:Country {name: 'Canada'}); // <2>
CREATE (alice)-[:LIVES_IN]->(:City {name: 'Toronto'}); // <3>

> CREATE (alice)-[:LIVES_IN {since: '1999-10-17'}]->(:City {name: 'Toronto'}) <4>

Reviewer: why does only some of the lines end with `;`?

(with `;` the MATCH is only applied to the first create that doesn't even use it XD I think for the relationship one we shouldn't have any `;` maybe)

----

<1> This will fail because the source node defined in the `LIVES_IN` relationship element type must have a `Resident` label.
<2> This will fail because the target node defined in the `LIVES_IN` relationship element type must have a `City` label.
<3> This will fail because an existence-constrained property (`since`) defined in the `LIVES_IN` relationship element type is missing.

> ④ This will fail because a type-constrained property (`since`) is of a different type (`STRING`) than the one defined for the property in the `LIVES_IN` relationship element type (`DATE`).

Reviewer: something is wrong with 4
<img width="1527" height="176" alt="image" src="https://github.com/user-attachments/assets/62e56798-8fc8-490d-8c87-315501a6efc1" />

```suggestion
<4> This will fail because a type-constrained property (`since`) is of a different type (`STRING`) than the one defined for the property in the `LIVES_IN` relationship element type (`DATE`).
```

[[open-graph-type-capabilities]]
== Open graph type capabilities

The great benefit of working with open graph types is that they allow you to constrain specific data while remaining flexible to add or modify data not constrained by the graph type.

.Create nodes and relationships with labels/types defined as identifying in the graph type, but with properties not defined in their respective node or relationship element types
[source, cypher]
----
CREATE (carl:Person:Resident {name: 'Carl Ericson', ssn: 162734679, born: date('1998-08-08')}), // <1>
       (molly:Pet:Resident:Animal {name: 'Molly', insuranceNumber: 672829172, healthCertificate: 'HT654987', breed: 'Alsatian'}), // <2>
       (atlanta:City {name: 'Atlanta', population: 2794356, country: 'USA'}), // <3>

>        (carl)-[:LIVES_IN {since: date('2020-09-15'), address: '101, Jasper Avenue'}]->(atlanta)<-[:LIVES_IN {since: date('2022-10-22')]-(molly) // <4>

Reviewer: ```suggestion
       (carl)-[:LIVES_IN {since: date('2020-09-15'), address: '101, Jasper Avenue'}]->(atlanta)<-[:LIVES_IN {since: date('2022-10-22')}]-(molly) // <4>
```

----

<1> There is no `born` property defined in the `Person` node element type.
<2> There is no `breed` property defined in the `Pet` node element type.
<3> There is no `country` property defined in the `City` node element type.
<4> There is no `address` property defined in the `LIVES_IN` relationship element type.

.Create nodes with implied labels but no identifying labels
[source, cypher]
----
CREATE (:StrayAnimal:Animal:Resident {id: '24.09-172898'}),
       (:Robot:Resident {name: 'Gary', application: 'Veterinary medicine'})
----

.Create nodes with node labels and relationship types not included in the graph type
[source, cypher]
----

> MATCH (ben:Person {name: 'Benjamin Davis}), (zoey:Pet {name: 'Zoey'}), (gary: Robot {name: 'Gary})

Reviewer: I noticed some weird colouring :P
<img width="1527" height="74" alt="image" src="https://github.com/user-attachments/assets/76291893-8d4b-4afa-8db3-9a4d5416511d" />

```suggestion
MATCH (ben:Person {name: 'Benjamin Davis'}), (zoey:Pet {name: 'Zoey'}), (gary:Robot {name: 'Gary'})
```

When situations like the example above arise, users should migrate the existing data to the new desired model and then apply the new graph type that will keep the data consistent from then on.

[[constraints-on-non-identifying-labels-and-types]]
=== Constraints on non-identifying node labels and relationship types

The examples thus far have included node and relationship element types, as well as property uniqueness and key constraints defined on identifying node labels and relationship types.

It is also possible to include constraints in a graph type on implied node labels, or on node labels and relationship types which are independent of a node or relationship element type, using the `CONSTRAINT … FOR … REQUIRE` syntax outlined xref:schema/graph-types/set-graph-types.adoc#naming-property-uniqueness-and-key-constraints[above].
This can be a useful tool if the data model requires certain attributes for nodes or relationships with non-identifying labels or types.

[NOTE]
Constraints on non-identifying node labels and relationship types can be explicitly named when defined.

> This can be useful when see xref:schema/graph-types/drop-graph-type-elements.adoc#drop-constraints-on-identifying-and-non-identifying-labels-and-types[Drop constraints on identifying and non-identifying node labels and relationship types] (if a name is not explicitly given, a name will be assigned).

Reviewer: ```suggestion
This can be useful when xref:schema/graph-types/drop-graph-type-elements.adoc#drop-constraints-on-identifying-and-non-identifying-labels-and-types[dropping constraints on identifying and non-identifying node labels and relationship types] (if a name is not explicitly given, a name will be assigned).
```
maybe? the `when see` felt weird at least (like either something was missing or extra)

.Set a graph type including property type constraint on non-identifying node label
[source, cypher]
----
ALTER CURRENT GRAPH TYPE SET {
(p:Person => :Resident {name :: STRING, ssn :: INTEGER})
   REQUIRE (p.name, p.ssn) IS KEY,
(:Pet => :Resident&Animal {insuranceNumber :: INTEGER IS KEY, healthCertificate :: STRING IS UNIQUE, name :: STRING}),
(:Robot => :Resident {application :: STRING NOT NULL}),
(:City => {name :: STRING NOT NULL, population :: INTEGER}),
(:Resident)-[:LIVES_IN => {since :: DATE NOT NULL}]->(:City),
CONSTRAINT company_name FOR (c:Company) REQUIRE c.name IS KEY,
CONSTRAINT animal_id FOR (a:Animal) REQUIRE a.id IS UNIQUE,
CONSTRAINT resident_address FOR (resident:Resident) REQUIRE resident.address IS :: STRING // <1>
}
----

<1> Property type constraint ensuring that any address properties on `RESIDENT` nodes are of type `STRING`.
This constraint is valid because `Resident` is not an identifying label in any of the above node element types.

[NOTE]
The example above sets a new graph type, *thus replacing the previously set graph type*.

> .Updated data model, showing new `Robot` node element type and addition of constraints on non-identifying node labels in graph type

Reviewer: that new Robot node element type just sneaks in XD we have one example where we try to add a non-valid robot type and then nothing about adding it until it's suddenly here XD (it can stay, I just didn't register it had sneaked in and got confused)

image::graph-types-set-graph-type-populated-database.svg[Graph type showing addition of constraints on non-identifying node labels, width=600, role=popup]

[[graph-type-implementation]]
== Graph type implementation: translation to constraints

Once a graph type is defined, it is translated into a set of constraints.
These constraints ensure that the creation and development of a graph conforms to the open schema specified by the graph type.

The constraints that enforce a graph type are returned by the `SHOW CONSTRAINTS` command.
For more information, see xref:schema/graph-types/list-graph-types.adoc#graph-type-elements-in-show-constraints[Show graph types -> Graph type elements in `SHOW CONSTRAINTS`].

.Set graph type
[source]
----

> Current graph type

Reviewer: <img width="806" height="212" alt="image" src="https://github.com/user-attachments/assets/e0ceb5ea-9467-4425-9fd1-1b6a69353db1" />
why not just have `Current graph type` as the header directly?

*Property existence constraints* ensure a property exists on all nodes with a specific label or all relationships with a specific type.
To learn more about property existence constraints and how to create them using the `CREATE CONSTRAINT` command, see xref:schema/constraints/create-constraints.adoc#create-property-existence-constraints[Constraints -> Create property existence constraints].

*Property uniqueness constraints* ensure that the combined property values are unique for all nodes with a specific label or all relationships with a specific type.
To learn more about property uniqueness constraints, and how to create them using the `CREATE CONSTRAINT` command, see xref:schema/constraints/create-constraints.adoc#create-property-uniqueness-constraints[Constraints -> Create property uniqueness constraints].

*Key constraints* ensure that all properties exist and that the combined property values are unique for all nodes with a specific label or all relationships with a specific type.

> To learn more about key constraints, and how to create them using the `CREATE CONSTRAINT` command, xref:schema/constraints/create-constraints.adoc#create-key-constraints[Constraints -> Create key constraints].

Reviewer: ```suggestion
To learn more about key constraints, and how to create them using the `CREATE CONSTRAINT` command, see xref:schema/constraints/create-constraints.adoc#create-key-constraints[Constraints -> Create key constraints].
```

Graph types can include:

Graph types can include:

> 

Reviewer: ```suggestion
Graph types can include:

```
no need to add it twice :P

* xref:schema/graph-types/set-graph-types.adoc#node-element-types[Node element types]
* xref:schema/graph-types/set-graph-types.adoc#relationship-element-types[Relationship element types]
* xref:schema/graph-types/set-graph-types.adoc#constraints-on-identifying-labels-and-types[Property uniqueness and key constraints defined on identifying node labels or relationship types]

> * xref:schema/graph-types/set-graph-types.adoc#constraints-on-non-identifying-labels-and-types[property existence, property type, property uniqueness, and key constraints defined on non-identifying node labels or relationship types]

Reviewer: saw that all other points started with capital letter 🤷 
```suggestion
* xref:schema/graph-types/set-graph-types.adoc#constraints-on-non-identifying-labels-and-types[Property existence, property type, property uniqueness, and key constraints defined on non-identifying node labels or relationship types]
```

.Create nodes with node labels and relationship types not included in the graph type
[source, cypher]
----

> MATCH (ben:Person {name: 'Benjamin Davis'}), (zoey:Pet {name: 'Zoey'}), (gary:Robot {name: 'Gary'})CREATE (ben)-[:OWNER_OF {ownershipId: 'GTHD-985'}]->(zoey),

Reviewer: ```suggestion
MATCH (ben:Person {name: 'Benjamin Davis'}), (zoey:Pet {name: 'Zoey'}), (gary:Robot {name: 'Gary'})
CREATE (ben)-[:OWNER_OF {ownershipId: 'GTHD-985'}]->(zoey),
```
